File: FrontEnd/src/app/services/member.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import {Member, MemberInfo} from '../models';
import { Observable } from 'rxjs';
import { environment } from "../../environments/environment";
import {Experience} from "../models/experience";
import {Education} from "../models/education";
import {Skills} from "../models/skills";



@Injectable({
  providedIn: 'root'
})
export class MemberService {

  private readonly membersUrl: string;

  constructor(private http: HttpClient) {
    this.membersUrl = `${environment.baseUrl}/api`;

  }

  public findAllMembers(): Observable<Member[]> {
    return this.http.get<Member[]>(`${this.membersUrl}/members`);
  }

  public checkIfEmailExists(memberEmail: string): Observable<boolean> {
    return this.http.get<boolean>(`${this.membersUrl}/members/get-email/${memberEmail}`);
  }

  public getMemberInfoByEmail(memberEmail: string): Observable<MemberInfo> {
    return this.http.get<MemberInfo>(`${this.membersUrl}/members/getInfo/${memberEmail}`);
  }

  public registerMember(member: Member): Observable<Member> {
    return this.http.post<Member>(`${this.membersUrl}/register`, member);
  }

  public updateCV(text: string, email: string): Observable<string> {
    return this.http.post<string>(`${this.membersUrl}/members/resume/update`, text, {
      headers: new HttpHeaders()
        .set('email', email),
      responseType: "text" as "json"
    })
  }

  public updateExperience(email: string, text: string): Observable<any> {
    return this.http.post<Experience>(`${this.membersUrl}/member/setExperience`, text, {
      headers: new HttpHeaders()
        .set('email', email),
      responseType: "text" as "json"
    });
  }

  public updateEducation(email: string, text: string): Observable<any> {
    return this.http.post<Education>(`${this.membersUrl}/member/setEducation`, text, {
      headers: new HttpHeaders()
        .set('email', email),
      responseType: "text" as "json"
    });
  }

  public updateSkills(email: string, text: string): Observable<any> {
    return this.http.post<Skills>(`${this.membersUrl}/member/setSkills`, text, {
      headers: new HttpHeaders()
        .set('email', email),
      responseType: "text" as "json"
    });
  }

  public updateEmail(old_email: string, new_email: string): Observable<string> {
    return this.http.post<string>(`${this.membersUrl}/members/change/email`, new_email, {
      headers: new HttpHeaders()
        .set('old_email', old_email),
      responseType: "text" as "json"
    })
  }

  public updatePassword(email: string, password: string): Observable<any> {
    return this.http.post<string>(`${this.membersUrl}/members/change/password`, password, {
      headers: new HttpHeaders()
        .set('email', email),
      responseType: "text" as "json"
    })
  }

  public publicResume(email: string): Observable<any>{
    return this.http.put<any>(`${this.membersUrl}/member/resume_public`, email, {
      responseType: "text" as "json"
    })
  }

  public privateResume(email: string): Observable<any>{
    return this.http.put<any>(`${this.membersUrl}/member/resume_private`, email, {
      responseType: "text" as "json"
    })
  }

  public privateEduc(email: string): Observable<any> {
    return this.http.put<any>(`${this.membersUrl}/member/education_private`, email, {
      responseType: "text" as "json"
    })
  }

  public publicEduc(email: string): Observable<any> {
    return this.http.put<any>(`${this.membersUrl}/member/education_public`, email, {
      responseType: "text" as "json"
    })
  }

  public publicExp(email: string): Observable<any> {
    return this.http.put<any>(`${this.membersUrl}/member/experience_public`, email, {
      responseType: "text" as "json"
    })
  }

  public privateExp(email: string): Observable<any> {
    return this.http.put<any>(`${this.membersUrl}/member/experience_private`, email, {
      responseType: "text" as "json"
    })
  }

  public privateSkills(email: string): Observable<any> {
    return this.http.put<any>(`${this.membersUrl}/member/skills_private`, email, {
      responseType: "text" as "json"
    })
  }

  public publicSkills(email: string): Observable<any> {
    return this.http.put<any>(`${this.membersUrl}/member/skills_public`, email, {
      responseType: "text" as "json"
    })
  }

  public getResume(email: string): Observable<any> {
    return this.http.get<any>(`${this.membersUrl}/member/resume/${email}`)
  }

  public getExperience(email: string): Observable<any>{
    return this.http.get<any>(`${this.membersUrl}/member/experience/${email}`)
  }

  public getEducation(email: string): Observable<any>{
    return this.http.get<any>(`${this.membersUrl}/member/education/${email}`)
  }

  public getSkills(email: string): Observable<any>{
    return this.http.get<any>(`${this.membersUrl}/member/skills/${email}`)
  }

}
export interface Credentials{
  password: string;
  email:    string;
}

File: BackEnd/src/main/java/com/example/backend/services/MemberService.java
package com.example.backend.services;

import java.util.List;

import com.example.backend.models.Education;
import com.example.backend.models.Experience;
import com.example.backend.models.Member;
import com.example.backend.models.MemberInfo;
import com.example.backend.models.MemberPersonalDataDTO;
import com.example.backend.models.Resume;
import com.example.backend.models.Role;
import com.example.backend.models.Skills;

public interface MemberService {

    // save member into DB
    Member saveMember(Member member);

    // get list of members
    List<MemberInfo> getMembers();

    // get Member specified by email
    Member findMemberByEmail(String email);

    Boolean userWithEmailExists(String email);

    MemberInfo getSpecifiedMemberInfo(String email);

    // get list of members of the specified emails
    List<Member> getSpecifiedMembers(List<String> emails);

    MemberPersonalDataDTO getMemberPersonalData(String email);

    // save role into DB
    Role saveRole(Role role);

    // add role to user
    void addRole(String email, String roleName);

    // save resume, experience, education, skills into DB
    Resume saveResume(Resume resume);

    Experience saveExperience(Experience experience);

    Education saveEducation(Education education);

    Skills saveSkills(Skills skills);


    // update resume if already exists in DB
    void updateResume(String text, Long resume_id);

    void updateExperience(String text, Long resume_id);

    void updateEducation(String text, Long resume_id);

    void updateSkills(String text, Long resume_id);

    void addExperience(String email, Experience experience);

    void addEducation(String email, Education education);

    void addSkills(String email, Skills skills);

    // add resume to user
    void addResume(String email, Resume resume);

    void publicResume(Long id);
    void privateResume(Long id);

    void publicEduc(Long id);
    void privateEduc(Long id);

    void publicExp(Long id);
    void privateExp(Long id);

    void publicSkills(Long id);
    void privateSkills(Long id);

    // change email
    void updateMemberEmail(String old_email, String new_email);

    // change password
    String updateMemberPassword(String email, String new_password, String jwt);

}
